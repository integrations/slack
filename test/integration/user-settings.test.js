const request = require('supertest');
const nock = require('nock');

const { probot, slackbot, models } = require('.');
const fixtures = require('../fixtures');

describe('Integration: User settings', () => {
  let slackUser;
  beforeEach(async () => {
    const { SlackWorkspace, SlackUser } = models;
    const workspace = await SlackWorkspace.create({
      slackId: 'T0001',
      accessToken: 'xoxa-token',
    });
    slackUser = await SlackUser.create({
      slackId: 'U2147483697',
      slackWorkspaceId: workspace.id,
    });
  });
  test('unauthenticated user is prompted to authenticate first', async () => {
    const command = fixtures.slack.command({
      text: 'settings',
    });
    const res = await request(probot.server).post('/slack/command')
      .use(slackbot)
      .send(command)
      .expect(200);
    const { text } = res.body.attachments[0].actions[0];
    expect(text).toMatch('Connect GitHub account');
  });
  describe('authenticated user', () => {
    beforeEach(async () => {
      const { GitHubUser } = models;
      const githubUser = await GitHubUser.create({
        id: 2,
        accessToken: 'github-token',
      });
      await slackUser.update({
        githubId: githubUser.id,
      });
    });

    test('with no settings configured gets message stating that', async () => {
      const command = fixtures.slack.command({
        text: 'settings',
      });
      await request(probot.server).post('/slack/command')
        .send(command)
        .expect(200)
        .expect((res) => {
          expect(res.body).toMatchSnapshot();
        });
    });

    test('with both automatic unfurls configured and prompts muted sees both settings', async () => {
      await slackUser.update({
        settings: {
          unfurlPrivateResources: {
            12345: ['all'],
            54321: ['C12345'],
          },
          muteUnfurlPromptsIndefinitely: true,
        },
      });

      nock('https://api.github.com').get('/repositories/12345').reply(200, {
        id: 12345,
        full_name: 'atom/atom',
      });

      nock('https://api.github.com').get('/repositories/54321').reply(200, {
        id: 54321,
        full_name: 'kubernetes/kubernetes',
      });

      const command = fixtures.slack.command({
        text: 'settings',
      });
      await request(probot.server).post('/slack/command')
        .send(command)
        .expect(200)
        .expect((res) => {
          expect(res.body).toMatchSnapshot();
        });
    });

    test('with just auto unfurls configured sees just that setting', async () => {
      await slackUser.update({
        settings: {
          unfurlPrivateResources: {
            12345: ['all'],
            54321: ['C12345'],
          },
        },
      });
      nock('https://api.github.com').get('/repositories/12345').reply(200, {
        id: 12345,
        full_name: 'atom/atom',
      });

      nock('https://api.github.com').get('/repositories/54321').reply(200, {
        id: 54321,
        full_name: 'kubernetes/kubernetes',
      });

      const command = fixtures.slack.command({
        text: 'settings',
      });
      await request(probot.server).post('/slack/command')
        .send(command)
        .expect(200)
        .expect((res) => {
          expect(res.body).toMatchSnapshot();
        });
    });

    test('with just prompts muted sees just that setting', async () => {
      await slackUser.update({
        settings: {
          muteUnfurlPromptsIndefinitely: true,
        },
      });

      const command = fixtures.slack.command({
        text: 'settings',
      });
      await request(probot.server).post('/slack/command')
        .send(command)
        .expect(200)
        .expect((res) => {
          expect(res.body).toMatchSnapshot();
        });
    });

    test('with prompts muted temporarily sees setting showing time remaining', async () => {
      Date.now = jest.fn(() => new Date(Date.UTC(2018, 4, 4)).valueOf());
      await slackUser.update({
        settings: {
          muteUnfurlPromptsUntil: 1525398771,
        },
      });

      const command = fixtures.slack.command({
        text: 'settings',
      });
      await request(probot.server).post('/slack/command')
        .send(command)
        .expect(200)
        .expect((res) => {
          expect(res.body).toMatchSnapshot();
        });
    });

    test('with prompts muted temporarily doesn\'t see setting showing time remaining', async () => {
      Date.now = jest.fn(() => new Date(Date.UTC(2018, 4, 4)).valueOf());
      await slackUser.update({
        settings: {
          muteUnfurlPromptsUntil: 1522713600,
        },
      });

      const command = fixtures.slack.command({
        text: 'settings',
      });
      await request(probot.server).post('/slack/command')
        .send(command)
        .expect(200)
        .expect((res) => {
          expect(res.body).toMatchSnapshot();
        });
    });

    describe('with both settings configured', () => {
      beforeEach(async () => {
        await slackUser.update({
          settings: {
            unfurlPrivateResources: {
              12345: ['all'],
              54321: ['C12345'],
            },
            muteUnfurlPromptsIndefinitely: true,
          },
        });
      });
      test('selects repo for which to stop autounfurls and is presented with a follow up message', async () => {
        await request(probot.server).post('/slack/actions').send({
          payload: JSON.stringify(fixtures.slack.action.unfurlSettingsAutoGetSettingsForRepo()),
        })
          .expect(200)
          .expect((res) => {
            expect(res.body).toMatchSnapshot();
          });
      });
      test('confirms removal of autounfurl removes repo from settings and posts confirmation message to user', async () => {
        await request(probot.server).post('/slack/actions').send({
          payload: JSON.stringify(fixtures.slack.action.unfurlSettingsAutoRemoveRepo()),
        })
          .expect(200)
          .expect((res) => {
            expect(res.body).toMatchSnapshot();
          });
        await slackUser.reload();
        expect(slackUser.settings.unfurlPrivateResources).toEqual({
          54321: ['C12345'],
        });
      });
      test('clicks unmute button and sees confirmation message', async () => {
        await request(probot.server).post('/slack/actions').send({
          payload: JSON.stringify(fixtures.slack.action.unfurlSettingsUnmutePrompts()),
        })
          .expect(200)
          .expect((res) => {
            expect(res.body).toMatchSnapshot();
          });
        await slackUser.reload();
        expect(slackUser.settings.muteUnfurlPromptsIndefinitely).toBe(false);
      });
    });
  });
});
