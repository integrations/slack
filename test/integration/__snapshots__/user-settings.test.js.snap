// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration: User settings authenticated user with both automatic unfurls configured and prompts muted sees both settings 1`] = `
Object {
  "attachments": Array [
    Object {
      "actions": Array [
        Object {
          "name": "get-settings-for-repo",
          "options": Array [
            Object {
              "text": "atom/atom",
              "value": "12345|atom/atom",
            },
            Object {
              "text": "kubernetes/kubernetes",
              "value": "54321|kubernetes/kubernetes",
            },
          ],
          "text": "Select a repository",
          "type": "select",
        },
      ],
      "callback_id": "unfurl-settings-auto",
      "color": "#24292f",
      "text": "Links you post linking to 2 private repositories automatically get a rich preview. Select a repository to adjust its settings.",
      "title": "Automatic rich previews",
    },
    Object {
      "actions": Array [
        Object {
          "name": "unmute-prompts",
          "text": "Unmute",
          "type": "button",
        },
      ],
      "callback_id": "unfurl-settings",
      "color": "#24292f",
      "text": "Prompts to show a rich preview are muted indefinitely",
      "title": "Muted prompts to show rich preview",
    },
  ],
}
`;

exports[`Integration: User settings authenticated user with both automatic unfurls configured and prompts muted sees both settings 2`] = `
Set {
  "pending-command#13345224609.738474920.8088930838d88f008e0",
}
`;

exports[`Integration: User settings authenticated user with both automatic unfurls configured and prompts muted sees both settings 3`] = `
Map {
  "pending-command#13345224609.738474920.8088930838d88f008e0" => Object {
    "channel_id": "C2147483705",
    "channel_name": "test",
    "command": "/github",
    "enterprise_id": "E0001",
    "enterprise_name": "Globular%20Construct%20Inc",
    "response_url": "https://hooks.slack.com/commands/1234/5678",
    "team_domain": "example",
    "team_id": "T0001",
    "text": "settings",
    "token": "secret",
    "trigger_id": "13345224609.738474920.8088930838d88f008e0",
    "user_id": "U2147483697",
    "user_name": "Steve",
  },
}
`;

exports[`Integration: User settings authenticated user with both settings configured clicks unmute button and sees confirmation message 1`] = `
Object {
  "attachments": Array [
    Object {
      "color": "#24292f",
      "text": ":white_check_mark: Prompts to show a rich preview have been unmuted",
    },
  ],
}
`;

exports[`Integration: User settings authenticated user with both settings configured clicks unmute button and sees confirmation message 2`] = `Set {}`;

exports[`Integration: User settings authenticated user with both settings configured clicks unmute button and sees confirmation message 3`] = `Map {}`;

exports[`Integration: User settings authenticated user with both settings configured confirms removal of autounfurl removes repo from settings and posts confirmation message to user 1`] = `
Object {
  "attachments": Array [
    Object {
      "color": "#24292f",
      "mrkdwn_in": Array [
        "text",
      ],
      "text": ":white_check_mark: Links you post to \`integrations/snappydoo\` will no longer receive automatic rich previews.",
    },
  ],
}
`;

exports[`Integration: User settings authenticated user with both settings configured confirms removal of autounfurl removes repo from settings and posts confirmation message to user 2`] = `Set {}`;

exports[`Integration: User settings authenticated user with both settings configured confirms removal of autounfurl removes repo from settings and posts confirmation message to user 3`] = `Map {}`;

exports[`Integration: User settings authenticated user with both settings configured selects repo for which to stop autounfurls and is presented with a follow up message 1`] = `
Object {
  "attachments": Array [
    Object {
      "actions": Array [
        Object {
          "confirm": Object {
            "dismiss_text": "Cancel",
            "ok_text": "Yes",
            "text": "This will disable automatic rich previews for links you post linking to \`integrations/snappydoo\` in *all* channels",
            "title": "Are you sure?",
          },
          "name": "remove-repo",
          "style": "danger",
          "text": "Disable automatic rich previews",
          "type": "button",
          "value": "12345|integrations/snappydoo",
        },
      ],
      "callback_id": "unfurl-settings-auto",
      "color": "#24292f",
      "mrkdwn_in": Array [
        "text",
      ],
      "text": "Links you post linking to \`integrations/snappydoo\` automatically get a rich preview in all channels.",
      "title": "Automatic rich previews",
    },
  ],
  "replace_original": true,
}
`;

exports[`Integration: User settings authenticated user with both settings configured selects repo for which to stop autounfurls and is presented with a follow up message 2`] = `Set {}`;

exports[`Integration: User settings authenticated user with both settings configured selects repo for which to stop autounfurls and is presented with a follow up message 3`] = `Map {}`;

exports[`Integration: User settings authenticated user with just auto unfurls configured sees just that setting 1`] = `
Object {
  "attachments": Array [
    Object {
      "actions": Array [
        Object {
          "name": "get-settings-for-repo",
          "options": Array [
            Object {
              "text": "atom/atom",
              "value": "12345|atom/atom",
            },
            Object {
              "text": "kubernetes/kubernetes",
              "value": "54321|kubernetes/kubernetes",
            },
          ],
          "text": "Select a repository",
          "type": "select",
        },
      ],
      "callback_id": "unfurl-settings-auto",
      "color": "#24292f",
      "text": "Links you post linking to 2 private repositories automatically get a rich preview. Select a repository to adjust its settings.",
      "title": "Automatic rich previews",
    },
  ],
}
`;

exports[`Integration: User settings authenticated user with just auto unfurls configured sees just that setting 2`] = `
Set {
  "pending-command#13345224609.738474920.8088930838d88f008e0",
}
`;

exports[`Integration: User settings authenticated user with just auto unfurls configured sees just that setting 3`] = `
Map {
  "pending-command#13345224609.738474920.8088930838d88f008e0" => Object {
    "channel_id": "C2147483705",
    "channel_name": "test",
    "command": "/github",
    "enterprise_id": "E0001",
    "enterprise_name": "Globular%20Construct%20Inc",
    "response_url": "https://hooks.slack.com/commands/1234/5678",
    "team_domain": "example",
    "team_id": "T0001",
    "text": "settings",
    "token": "secret",
    "trigger_id": "13345224609.738474920.8088930838d88f008e0",
    "user_id": "U2147483697",
    "user_name": "Steve",
  },
}
`;

exports[`Integration: User settings authenticated user with just prompts muted sees just that setting 1`] = `
Object {
  "attachments": Array [
    Object {
      "actions": Array [
        Object {
          "name": "unmute-prompts",
          "text": "Unmute",
          "type": "button",
        },
      ],
      "callback_id": "unfurl-settings",
      "color": "#24292f",
      "text": "Prompts to show a rich preview are muted indefinitely",
      "title": "Muted prompts to show rich preview",
    },
  ],
}
`;

exports[`Integration: User settings authenticated user with just prompts muted sees just that setting 2`] = `
Set {
  "pending-command#13345224609.738474920.8088930838d88f008e0",
}
`;

exports[`Integration: User settings authenticated user with just prompts muted sees just that setting 3`] = `
Map {
  "pending-command#13345224609.738474920.8088930838d88f008e0" => Object {
    "channel_id": "C2147483705",
    "channel_name": "test",
    "command": "/github",
    "enterprise_id": "E0001",
    "enterprise_name": "Globular%20Construct%20Inc",
    "response_url": "https://hooks.slack.com/commands/1234/5678",
    "team_domain": "example",
    "team_id": "T0001",
    "text": "settings",
    "token": "secret",
    "trigger_id": "13345224609.738474920.8088930838d88f008e0",
    "user_id": "U2147483697",
    "user_name": "Steve",
  },
}
`;

exports[`Integration: User settings authenticated user with no settings configured gets message stating that 1`] = `
Object {
  "attachments": Array [
    Object {
      "color": "#24292f",
      "text": "You do not have any settings configured.",
    },
  ],
}
`;

exports[`Integration: User settings authenticated user with no settings configured gets message stating that 2`] = `
Set {
  "pending-command#13345224609.738474920.8088930838d88f008e0",
}
`;

exports[`Integration: User settings authenticated user with no settings configured gets message stating that 3`] = `
Map {
  "pending-command#13345224609.738474920.8088930838d88f008e0" => Object {
    "channel_id": "C2147483705",
    "channel_name": "test",
    "command": "/github",
    "enterprise_id": "E0001",
    "enterprise_name": "Globular%20Construct%20Inc",
    "response_url": "https://hooks.slack.com/commands/1234/5678",
    "team_domain": "example",
    "team_id": "T0001",
    "text": "settings",
    "token": "secret",
    "trigger_id": "13345224609.738474920.8088930838d88f008e0",
    "user_id": "U2147483697",
    "user_name": "Steve",
  },
}
`;

exports[`Integration: User settings authenticated user with prompts muted temporarily doesn't see setting showing time remaining 1`] = `
Object {
  "attachments": Array [
    Object {
      "color": "#24292f",
      "text": "You do not have any settings configured.",
    },
  ],
}
`;

exports[`Integration: User settings authenticated user with prompts muted temporarily doesn't see setting showing time remaining 2`] = `
Set {
  "pending-command#13345224609.738474920.8088930838d88f008e0",
}
`;

exports[`Integration: User settings authenticated user with prompts muted temporarily doesn't see setting showing time remaining 3`] = `
Map {
  "pending-command#13345224609.738474920.8088930838d88f008e0" => Object {
    "channel_id": "C2147483705",
    "channel_name": "test",
    "command": "/github",
    "enterprise_id": "E0001",
    "enterprise_name": "Globular%20Construct%20Inc",
    "response_url": "https://hooks.slack.com/commands/1234/5678",
    "team_domain": "example",
    "team_id": "T0001",
    "text": "settings",
    "token": "secret",
    "trigger_id": "13345224609.738474920.8088930838d88f008e0",
    "user_id": "U2147483697",
    "user_name": "Steve",
  },
}
`;

exports[`Integration: User settings authenticated user with prompts muted temporarily sees setting showing time remaining 1`] = `
Object {
  "attachments": Array [
    Object {
      "actions": Array [
        Object {
          "name": "unmute-prompts",
          "text": "Unmute",
          "type": "button",
        },
      ],
      "callback_id": "unfurl-settings",
      "color": "#24292f",
      "text": "Prompts to show a rich preview are muted for 2 hours",
      "title": "Muted prompts to show rich preview",
    },
  ],
}
`;

exports[`Integration: User settings authenticated user with prompts muted temporarily sees setting showing time remaining 2`] = `
Set {
  "pending-command#13345224609.738474920.8088930838d88f008e0",
}
`;

exports[`Integration: User settings authenticated user with prompts muted temporarily sees setting showing time remaining 3`] = `
Map {
  "pending-command#13345224609.738474920.8088930838d88f008e0" => Object {
    "channel_id": "C2147483705",
    "channel_name": "test",
    "command": "/github",
    "enterprise_id": "E0001",
    "enterprise_name": "Globular%20Construct%20Inc",
    "response_url": "https://hooks.slack.com/commands/1234/5678",
    "team_domain": "example",
    "team_id": "T0001",
    "text": "settings",
    "token": "secret",
    "trigger_id": "13345224609.738474920.8088930838d88f008e0",
    "user_id": "U2147483697",
    "user_name": "Steve",
  },
}
`;

exports[`Integration: User settings unauthenticated user is prompted to authenticate first 1`] = `
Set {
  "pending-command#13345224609.738474920.8088930838d88f008e0",
}
`;

exports[`Integration: User settings unauthenticated user is prompted to authenticate first 2`] = `
Map {
  "pending-command#13345224609.738474920.8088930838d88f008e0" => Object {
    "channel_id": "C2147483705",
    "channel_name": "test",
    "command": "/github",
    "enterprise_id": "E0001",
    "enterprise_name": "Globular%20Construct%20Inc",
    "response_url": "https://hooks.slack.com/commands/1234/5678",
    "team_domain": "example",
    "team_id": "T0001",
    "text": "settings",
    "token": "secret",
    "trigger_id": "13345224609.738474920.8088930838d88f008e0",
    "user_id": "U2147483697",
    "user_name": "Steve",
  },
}
`;
